src/doc.cpp

```diff
@@ -1596,4 +1598,172 @@ int Doc::PrepareTimestampsEnd(FunctorParams *functorParams)
+            // @next
+            oldIdString = interface->GetNext();
+            if (oldIdString.rfind("#", 0) == 0) oldIdString = oldIdString.substr(1, oldIdString.size() - 1);
+            newIdString = getExpansionIdsForElement(oldIdString).back();
+            if (!newIdString.empty()) interface->SetNext("#" + newIdString);
+            // @next
+            oldIdString = interface->GetPrev();
+            if (oldIdString.rfind("#", 0) == 0) oldIdString = oldIdString.substr(1, oldIdString.size() - 1);
+            newIdString = getExpansionIdsForElement(oldIdString).back();
+            if (!newIdString.empty()) interface->SetNext("#" + newIdString);
+            // @prev
+            oldIdString = interface->GetPrev();
+            if (oldIdString.rfind("#", 0) == 0) oldIdString = oldIdString.substr(1, oldIdString.size() - 1);
+            newIdString = getExpansionIdsForElement(oldIdString).back();
+            if (!newIdString.empty()) interface->SetNext("#" + newIdString);
+            // @copyof
+            oldIdString = interface->GetCopyof();
+            if (oldIdString.rfind("#", 0) == 0) oldIdString = oldIdString.substr(1, oldIdString.size() - 1);
+            newIdString = getExpansionIdsForElement(oldIdString).back();
+            if (!newIdString.empty()) interface->SetCopyof("#" + newIdString);
+            // @corresp
+            oldIdString = interface->GetCorresp();
+            if (oldIdString.rfind("#", 0) == 0) oldIdString = oldIdString.substr(1, oldIdString.size() - 1);
+            newIdString = getExpansionIdsForElement(oldIdString).back();
+            if (!newIdString.empty()) interface->SetCorresp("#" + newIdString);
+            // @synch
+            oldIdString = interface->GetSynch();
+            if (oldIdString.rfind("#", 0) == 0) oldIdString = oldIdString.substr(1, oldIdString.size() - 1);
+            newIdString = getExpansionIdsForElement(oldIdString).back();
+            if (!newIdString.empty()) interface->SetSynch("#" + newIdString);
+        }
+        updateIds(o);
+    }
+    return true;
+}
+
+bool Doc::addExpandedIdToExpansionMap(const std::string &origXmlId, std::string newXmlId)
+{
+    bool found = false;
+    std::vector<std::vector<std::string> >::iterator iter;
+    for (iter = m_expansionMap.begin(); iter != m_expansionMap.end(); ++iter) {
+        auto it = std::find(iter->begin(), iter->end(), origXmlId);
+        if (it != iter->end()) {
+            iter->push_back(newXmlId);
+            found = true;
+            break;
+        }
+    }
+    if (!found) {
+        std::vector<std::string> ids;
+        ids.push_back(origXmlId.c_str());
+        ids.push_back(newXmlId.c_str());
+        m_expansionMap.push_back(ids);
+    }
+    return true;
+}
+
+std::vector<std::string> Doc::getExpansionIdsForElement(const std::string &xmlId)
+{
+    std::vector<std::vector<std::string> >::iterator iter;
+    for (iter = m_expansionMap.begin(); iter != m_expansionMap.end(); ++iter) {
```

Comment:
This lookup would not be necessary if you use a std::map instead

Reply:
That's true; see above.